_posts/2020-10-19-torch-image-classification/torch-image-classification.Rmd

```diff
@@ -0,0 +1,543 @@
+```
+
+#### Finding an optimally efficient learning rate
+
+We set the learning rate to `0.1`, but that is just a formality. As became widely known due to the excellent lectures by [fast.ai](http://fast.ai), it makes sense to spend some time upfront to determine an efficient learning rate. While out-of-the-box, `torch` does not provide a tool like fast.ai's learning rate finder, the logic is straightforward to implement. Here's how to find a good learning rate, as translated to R from [Sylvain Gugger's post](https://sgugger.github.io/how-do-you-find-a-good-learning-rate.html):
+
+```{r}
+# ported from: https://sgugger.github.io/how-do-you-find-a-good-learning-rate.html
+
+losses <- c()
+log_lrs <- c()
+
+find_lr <- function(init_value = 1e-8, final_value = 10, beta = 0.98) {
+
+  num <- train_dl$.length()
+  mult = (final_value/init_value)^(1/num)
+  lr <- init_value
+  optimizer$param_groups[[1]]$lr <- lr
+  avg_loss <- 0
+  best_loss <- 0
+  batch_num <- 0
+
+  for (b in enumerate(train_dl)) {
+
+    batch_num <- batch_num + 1
+    optimizer$zero_grad()
+    output <- model(b[[1]]$to(device = device))
+    loss <- criterion(output, b[[2]]$to(device = device))
+
+    #Compute the smoothed loss
+    avg_loss <- beta * avg_loss + (1-beta) * loss$item()
+    smoothed_loss <- avg_loss / (1 - beta^batch_num)
+    #Stop if the loss is exploding
+    if (batch_num > 1 && smoothed_loss > 4 * best_loss) break
+    #Record the best loss
+    if (smoothed_loss < best_loss || batch_num == 1) best_loss <- smoothed_loss
+
+    #Store the values
+    losses <<- c(losses, smoothed_loss)
+    log_lrs <<- c(log_lrs, (log(lr, 10)))
+
+    loss$backward()
+    optimizer$step()
+
+    #Update the lr for the next step
+    lr <- lr * mult
+    optimizer$param_groups[[1]]$lr <- lr
+  }
+}
+
+find_lr()
+
+df <- data.frame(log_lrs = log_lrs, losses = losses)
+ggplot(df, aes(log_lrs, losses)) + geom_point(size = 1) + theme_classic()
+
+```
+
+```{r, eval = TRUE, echo = FALSE, layout="l-body-outset"}
+knitr::include_graphics("images/lr_finder.png")
+```
+
+The best learning rate is not the exact one where loss is at a minimum. Instead, it should be picked somewhat earlier on the curve, while loss is still decreasing. `0.05` looks like a sensible choice.
+
+This value is nothing but an anchor, however. *Learning rate schedulers* allow to evolve learning rates according to some proven algorithm. Among others, `torch` implements one-cycle learning [\@abs-1708-07120], cyclical learning rates [@Smith15a], and cosine annealing with warm restarts [@LoshchilovH16a].
+
+Here, we use `lr_one_cycle()`, passing in our newly found, optimally efficient, hopefully, value `0.05` as a maximum learning rate. `lr_one_cycle()` will start with a low rate, then gradually ramp up until it reaches the allowed maximum. After that, the learning rate will slowly, continually decrease, until it falls slightly below its initial value.
```

Comment:
```suggestion
Here, we use `lr_one_cycle()`, passing in our newly found, optimally efficient (hopefully) value `0.05` as a maximum learning rate. `lr_one_cycle()` will start with a low rate, then gradually ramp up until it reaches the allowed maximum. After that, the learning rate will slowly decrease continually, until it falls slightly below its initial value.
```

Reply:
I think you've got continually right here, but I _always_ mix up continuously and continually, should you want to double check: https://www.grammarly.com/blog/continuously-continually/#:~:text=The%20adverbs%20continuously%20and%20continually,that%20recurs%20frequently%20or%20regularly.

Reply:
oh wow that's interesting! I didn't even google it because it stayed half-subconscious somehow ... but this means I chose the wrong one in fact